File: NostrVault/app/src/main/java/com/nostrvault/service/ReelsFeedService.kt
package com.nostrvault.service

import android.content.Context
import android.util.Log
import com.nostrvault.data.local.ConfigStore
import com.nostrvault.data.model.FeedNote
import com.nostrvault.data.model.Reel
import com.nostrvault.data.model.ReelCursors
import com.nostrvault.data.model.ReelStream
import com.nostrvault.data.model.ReelsScope
import com.nostrvault.data.remote.WebSocketClient
import com.nostrvault.relay.RelayForegroundService
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Reels: one playable video per page, swiped vertically. Port of iOS
 * `ReelsFeedService`, shaped like [LiveFeedService]: its own short-lived relay
 * connections, never the note subscription.
 *
 * Two sources feed it. NIP-71 video events (kinds 21/22 and their addressable
 * 34235/34236 forms) are video by definition. Ordinary kind-1 notes are where
 * most video on Nostr actually lives, so they are fetched too and kept only
 * when a URL in them is known to be video — see [Reel.from].
 *
 * Threading: relay messages are parsed on a background dispatcher, one
 * collector per relay, and every piece of paging state is read and written
 * under [lock]. Hopping each event to the main thread instead would let a
 * relay's burst overrun [WebSocketClient]'s message buffer, which drops rather
 * than waits. Every fetch carries a generation number, so a message or timeout
 * from a torn-down page can never touch the current one.
 */
@Singleton
class ReelsFeedService @Inject constructor(
    @ApplicationContext context: Context,
    private val configStore: ConfigStore,
    private val nostrService: NostrService,
    private val feedService: FeedService,
    private val mediaTypeDetector: MediaTypeDetector,
) {
    companion object {
        private const val TAG = "ReelsFeedService"

        /**
         * How long one page may wait on relays that have not answered. A page
         * every relay has answered finishes at once; this only bounds a dead
         * one. Longer than iOS's 10s: on the phone, five sockets doing DNS, TLS
         * and a REQ were measured connecting as late as t+5s (see
         * [LiveFeedService.COLLECT_WINDOW_MS]'s note).
         */
        private const val PAGE_TIMEOUT_MS = 15_000L

        /** Events arrive one message at a time from several relays; batch them. */
        private const val PUBLISH_DEBOUNCE_MS = 350L

        private const val PREFS_NAME = "reels_preferences"
        private const val KEY_MUTED = "reels_muted"
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val json = Json { ignoreUnknownKeys = true }
    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    // ── Observable state ──────────────────────────────────────────────

    private val _reels = MutableStateFlow<List<Reel>>(emptyList())
    /** Playable reels in play order. */
    val reels: StateFlow<List<Reel>> = _reels.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    /** The first page is still being fetched. */
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isLoadingMore = MutableStateFlow(false)
    val isLoadingMore: StateFlow<Boolean> = _isLoadingMore.asStateFlow()

    private val _loadFailed = MutableStateFlow(false)
    val loadFailed: StateFlow<Boolean> = _loadFailed.asStateFlow()

    private val _followSetIsEmpty = MutableStateFlow(false)
    val followSetIsEmpty: StateFlow<Boolean> = _followSetIsEmpty.asStateFlow()

    private val _scope = MutableStateFlow(ReelsScope.FOLLOWING)
    /**
     * Following by default. Global video is unmoderated third-party content,
     * so it is opt-in behind the sensitive-content warning.
     */
    val reelsScope: StateFlow<ReelsScope> = _scope.asStateFlow()

    private val _isMuted = MutableStateFlow(prefs.getBoolean(KEY_MUTED, false))
    /** Sound follows the viewer from reel to reel and across launches. */
    val isMuted: StateFlow<Boolean> = _isMuted.asStateFlow()

    // ── Paging state (guarded by lock) ────────────────────────────────

    private val lock = Any()
    private var generation = 0
    private val clients = mutableListOf<WebSocketClient>()
    private val clientJobs = mutableListOf<Job>()
    private var timeoutJob: Job? = null
    private var publishJob: Job? = null

    private val collected = HashMap<String, Reel>()
    private val seenVideoUrls = HashSet<String>()
    /**
     * Once the viewer has swiped past the first reel the order is frozen:
     * arrivals are appended instead of sorted in, so the video under their
     * thumb never jumps.
     */
    private var orderIsFrozen = false
    private var shownReelId: String? = null

    private var cursors = ReelCursors()
    /** Per relay index, per stream: the oldest event this page returned. */
    private val pageOldest = HashMap<Int, MutableMap<ReelStream, Long>>()
    /** Relays that have finished this page (EOSE or CLOSED). */
    private val answered = HashSet<Int>()
    private var countBeforeFetch = 0

    private val isFetching: Boolean get() = clients.isNotEmpty()

    // ── Public API ────────────────────────────────────────────────────

    fun setScope(newScope: ReelsScope) {
        synchronized(lock) {
            if (newScope == _scope.value) return
            _scope.value = newScope
            refreshLocked()
        }
    }

    fun setMuted(muted: Boolean) {
        _isMuted.value = muted
        prefs.edit().putBoolean(KEY_MUTED, muted).apply()
    }

    fun loadIfNeeded() {
        synchronized(lock) {
            if (isFetching || _reels.value.isNotEmpty()) return
            refreshLocked()
        }
    }

    fun refresh() {
        synchronized(lock) { refreshLocked() }
    }

    /**
     * Called as the viewer nears the end of what is loaded. A page still in
     * flight is left alone — starting another would cut off the relays that
     * have not answered yet.
     */
    fun loadMore() {
        synchronized(lock) {
            if (isFetching || cursors.reachedEnd || _reels.value.isEmpty()) return
            _isLoadingMore.value = true
            fetchLocked()
        }
    }

    /** The view reports the reel on screen; leaving the first one freezes order. */
    fun didShow(reelId: String) {
        synchronized(lock) {
            shownReelId = reelId
            val first = _reels.value.firstOrNull()
            if (!orderIsFrozen && first != null && first.id != reelId) orderIsFrozen = true
        }
    }

    /** Stops the page in flight. Leaving the mode calls this. */
    fun disconnect() {
        synchronized(lock) {
            disconnectLocked()
            _isLoading.value = false
            _isLoadingMore.value = false
        }
    }

    /** Drops reels by anyone blocked since they were fetched. */
    fun pruneBlocked() {
        synchronized(lock) {
            val blocked = blockedHex()
            if (blocked.isEmpty() || _reels.value.none { it.note.pubkey in blocked }) return
            collected.values.removeAll { it.note.pubkey in blocked }
            _reels.value = _reels.value.filter { it.note.pubkey !in blocked }
        }
    }

    // ── Fetching (call with lock held) ────────────────────────────────

    private fun refreshLocked() {
        disconnectLocked()
        collected.clear()
        seenVideoUrls.clear()
        _reels.value = emptyList()
        orderIsFrozen = false
        shownReelId = null
        cursors = ReelCursors()
        _loadFailed.value = false
        _followSetIsEmpty.value = false
        _isLoadingMore.value = false
        _isLoading.value = true
        fetchLocked()
    }

    private fun disconnectLocked() {
        // Bumping the generation is what makes late messages and timers from
        // the old page harmless; cancelling is what stops the work.
        generation++
        timeoutJob?.cancel()
        timeoutJob = null
        publishJob?.cancel()
        publishJob = null
        clientJobs.forEach { it.cancel() }
        clientJobs.clear()
        clients.forEach { it.disconnect() }
        clients.clear()
    }

    private fun fetchLocked() {
        disconnectLocked()
        val gen = generation

        val authors: List<String>? = if (_scope.value == ReelsScope.FOLLOWING) {
            val follows = feedService.followedPubkeys.value
            if (follows.isEmpty()) {
                _isLoading.value = false
                _isLoadingMore.value = false
                _followSetIsEmpty.value = true
                return
            }
            follows
        } else {
            null
        }

        val filters = cursors.filters(authors)
        if (filters.isEmpty()) {
            _isLoading.value = false
            _isLoadingMore.value = false
            return
        }

        val relayUrls = relayUrls(_scope.value)
        if (relayUrls.isEmpty()) {
            _isLoading.value = false
            _isLoadingMore.value = false
            _loadFailed.value = true
            return
        }

        val subId = "reels-${System.nanoTime().toString(36).takeLast(8)}"
        val req = "[\"REQ\",\"$subId\",${filters.joinToString(",")}]"
        val blocked = blockedHex()
        countBeforeFetch = collected.size
        answered.clear()
        pageOldest.clear()

        for ((relayIndex, url) in relayUrls.withIndex()) {
            val client = WebSocketClient(
                url = url,
                scope = scope,
                trustLocalhost = url.contains("localhost") || url.contains("127.0.0.1"),
            )
            clients += client
            clientJobs += scope.launch(Dispatchers.Default, start = CoroutineStart.UNDISPATCHED) {
                client.messages.collect { raw -> onMessage(raw, subId, relayIndex, gen, blocked) }
            }
            clientJobs += scope.launch {
                client.connectionState.collect { state ->
                    if (state == WebSocketClient.ConnectionState.CONNECTED) client.send(req)
                }
            }
            client.connect()
        }

        timeoutJob = scope.launch {
            delay(PAGE_TIMEOUT_MS)
            synchronized(lock) { finishPageLocked(gen) }
        }
    }

    /**
     * The device's own relay (and its feed cache) answer first and hold the
     * follow set's history; the feed relays fill in the rest.
     */
    private fun relayUrls(scope: ReelsScope): List<String> {
        val config = configStore.config.value
        return buildList {
            if (scope == ReelsScope.FOLLOWING && RelayForegroundService.readyForConnections.value) {
                config.nostrURL?.let(::add)
                config.localRelayURL("feed")?.let(::add)
            }
            addAll(config.activeFeedRelays.ifEmpty { listOf("wss://relay.primal.net", "wss://nos.lol") })
        }.distinct()
    }

    private fun blockedHex(): Set<String> = configStore.config.value.blockedForActiveAccount()
        .mapNotNull { nostrService.npubToHex(it) }
        .toSet()

    // ── Messages (background dispatcher) ──────────────────────────────

    private fun onMessage(raw: String, subId: String, relay: Int, gen: Int, blocked: Set<String>) {
        val array = try {
            json.parseToJsonElement(raw) as? JsonArray
        } catch (e: Exception) {
            Log.w(TAG, "unparseable relay message: ${e.message}")
            null
        } ?: return
        if (array.size < 2) return
        val type = array[0].jsonPrimitive.contentOrNull
        if (array[1].jsonPrimitive.contentOrNull != subId) return

        when (type) {
            "EOSE", "CLOSED" -> synchronized(lock) {
                if (gen != generation) return
                answered += relay
                // Every relay has answered — no reason to sit out the timeout.
                if (answered.size >= clients.size) finishPageLocked(gen) else schedulePublishLocked(immediate = true)
            }
            "EVENT" -> {
                val event = array.getOrNull(2) as? JsonObject ?: return
                val reel = parseEvent(event, blocked) ?: return
                synchronized(lock) {
                    if (gen != generation) return
                    val oldest = pageOldest.getOrPut(relay) { mutableMapOf() }
                    oldest[reel.stream] = minOf(oldest[reel.stream] ?: reel.createdAt, reel.createdAt)

                    val candidate = reel.reel ?: return
                    if (collected.containsKey(candidate.id)) return
                    // The same clip re-posted, or a NIP-71 event and the note announcing it.
                    if (!seenVideoUrls.add(candidate.videoUrl)) return
                    collected[candidate.id] = candidate
                    schedulePublishLocked(immediate = false)
                }
            }
        }
    }

    /** What one EVENT contributes: its paging position always, a reel when it has one. */
    private class ParsedEvent(val stream: ReelStream, val createdAt: Long, val reel: Reel?)

    private fun parseEvent(event: JsonObject, blocked: Set<String>): ParsedEvent? = try {
        val kind = event["kind"]?.jsonPrimitive?.contentOrNull?.toIntOrNull()
        val stream = kind?.let(ReelStream::forKind)
        val id = event["id"]?.jsonPrimitive?.contentOrNull
        val pubkey = event["pubkey"]?.jsonPrimitive?.contentOrNull
        val createdAt = event["created_at"]?.jsonPrimitive?.contentOrNull?.toLongOrNull()
        if (stream == null || id == null || pubkey == null || createdAt == null) {
            null
        } else {
            // A blocked author's event still marks how far this relay got.
            val reel = if (pubkey in blocked) {
                null
            } else {
                val tags = event["tags"]?.jsonArray?.map { tag ->
                    tag.jsonArray.map { it.jsonPrimitive.contentOrNull.orEmpty() }
                } ?: emptyList()
                val note = FeedNote.fromEvent(
                    id = id,
                    pubkey = pubkey,
                    content = event["content"]?.jsonPrimitive?.contentOrNull.orEmpty(),
                    tags = tags,
                    createdAt = createdAt,
                    kind = kind!!,
                )
                Reel.from(note, createdAt, mediaTypeDetector::getCachedContentType)
            }
            ParsedEvent(stream, createdAt, reel)
        }
    } catch (e: Exception) {
        Log.w(TAG, "unusable event: ${e.message}")
        null
    }

    // ── Publishing (call with lock held) ──────────────────────────────

    private fun schedulePublishLocked(immediate: Boolean) {
        publishJob?.cancel()
        val gen = generation
        publishJob = scope.launch(Dispatchers.Default) {
            if (!immediate) delay(PUBLISH_DEBOUNCE_MS)
            synchronized(lock) { if (gen == generation) publishLocked() }
        }
    }

    private fun publishLocked() {
        val current = _reels.value
        if (orderIsFrozen) {
            val shown = current.mapTo(HashSet()) { it.id }
            val arrivals = collected.values.filter { it.id !in shown }.sortedWith(Reel.NEWEST_FIRST)
            if (arrivals.isNotEmpty()) _reels.value = current + arrivals
        } else {
            _reels.value = collected.values.sortedWith(Reel.NEWEST_FIRST)
        }
    }

    private fun finishPageLocked(gen: Int) {
        if (gen != generation || !isFetching) return
        publishLocked()

        // No relay answered at all: a network failure, not the end of
        // history. Leave the cursors alone so a later swipe can retry.
        if (pageOldest.isEmpty() && answered.isEmpty()) {
            disconnectLocked()
            _isLoading.value = false
            _isLoadingMore.value = false
            _loadFailed.value = _reels.value.isEmpty() && !_followSetIsEmpty.value
            return
        }

        cursors = cursors.afterPage(pageOldest.values, producedReels = collected.size > countBeforeFetch)
        disconnectLocked()
        _isLoading.value = false
        _isLoadingMore.value = false
        _loadFailed.value = false

        // Keep paging while the viewer is already near the end — nothing else
        // would trigger the next page once the pager sits on its last reel.
        val reels = _reels.value
        val shownIndex = shownReelId?.let { id -> reels.indexOfFirst { it.id == id } }?.takeIf { it >= 0 } ?: 0
        if (ReelCursors.shouldContinue(reels.size, shownIndex, cursors.reachedEnd)) {
            if (reels.isEmpty()) _isLoading.value = true else _isLoadingMore.value = true
            fetchLocked()
        }
    }
}

File: HavenApp/HavenApp/Services/ReelsFeedService.swift
import Foundation
import Combine

/// Reels: one playable video per page, swiped vertically.
///
/// Two sources feed it. NIP-71 video events (kinds 21/22 and their addressable
/// 34235/34236 forms) are video by definition. Ordinary kind-1 notes are where
/// most video on Nostr actually lives, so they are fetched too and kept only
/// when a URL in them is known to be video — by extension or by the `m` field
/// of its NIP-92 `imeta` tag. Extensionless links with no MIME hint are left
/// out rather than HEAD-requested one by one: a page that turns out to be an
/// image is worse than a missing one.
@MainActor
final class ReelsFeedService: ObservableObject {
    static let shared = ReelsFeedService()

    /// Playable reels in play order.
    @Published private(set) var reels: [Reel] = []
    /// The first page is still being fetched.
    @Published private(set) var isLoading = false
    @Published private(set) var isLoadingMore = false
    @Published private(set) var loadFailed = false
    @Published private(set) var followSetIsEmpty = false
    /// Following by default. Global video is unmoderated third-party content,
    /// so it is opt-in behind the sensitive-content warning, like Media.
    @Published private(set) var scope: RecipeScope = .following

    static let videoKinds = [21, 22, 34235, 34236]

    /// The two filters a page asks for. Each pages on its own cursor: kind-1
    /// notes are dense and NIP-71 events are sparse, so one shared cursor would
    /// either skip notes or re-ask for the same video window.
    private enum Stream: CaseIterable { case video, note }

    private var clients: [WebSocketClient] = []
    private var cancellables = Set<AnyCancellable>()
    private var collected: [String: Reel] = [:]
    private var seenVideoURLs: Set<String> = []
    private var loadTimeout: Timer?
    private var publishWork: DispatchWorkItem?
    /// Once the viewer has swiped past the first reel the order is frozen:
    /// arrivals are appended instead of sorted in, so the video under their
    /// thumb never jumps.
    private var orderIsFrozen = false
    private var shownReelId: String?

    /// Where the next page starts, per stream. A stream missing from the map
    /// has run out of history.
    private var cursors: [Stream: Int64] = [:]
    private var exhausted: Set<Stream> = []
    /// Per relay, per stream: the oldest event this page returned.
    private var pageOldest: [Int: [Stream: Int64]] = [:]
    private var eoseThisFetch = 0
    private var countBeforeFetch = 0
    /// Consecutive pages that produced no reel; stops a video-less stretch of
    /// history from paging forever.
    private var emptyPages = 0
    private var isFetching: Bool { !clients.isEmpty }
    private var reachedEnd: Bool { exhausted.count == Stream.allCases.count || emptyPages >= 4 }

    private init() {}

    func setScope(_ newScope: RecipeScope) {
        guard newScope != scope else { return }
        scope = newScope
        refresh()
    }

    func loadIfNeeded() {
        guard !isFetching, reels.isEmpty else { return }
        refresh()
    }

    func refresh() {
        disconnect()
        collected.removeAll()
        seenVideoURLs.removeAll()
        reels = []
        orderIsFrozen = false
        shownReelId = nil
        cursors.removeAll()
        exhausted.removeAll()
        emptyPages = 0
        loadFailed = false
        followSetIsEmpty = false
        isLoadingMore = false
        isLoading = true
        fetch()
    }

    /// Called by the view as the viewer nears the end of what is loaded. A
    /// page still in flight is left alone — starting another would cut off
    /// the relays that have not answered yet.
    func loadMore() {
        guard !isFetching, !reachedEnd, !reels.isEmpty else { return }
        isLoadingMore = true
        fetch()
    }

    /// The view reports the reel on screen; leaving the first one freezes order.
    func didShow(reelId: String) {
        shownReelId = reelId
        if !orderIsFrozen, let first = reels.first, first.id != reelId {
            orderIsFrozen = true
        }
    }

    func disconnect() {
        loadTimeout?.invalidate()
        loadTimeout = nil
        publishWork?.cancel()
        publishWork = nil
        cancellables.removeAll()
        clients.forEach { $0.disconnect() }
        clients.removeAll()
    }

    /// Drops reels by anyone blocked since they were fetched. Cheap enough to
    /// run whenever a sheet that can block (a profile) closes.
    func pruneBlocked() {
        let blocked = ConfigService.shared.activeAccountBlockedHexPubkeys
        guard !blocked.isEmpty, reels.contains(where: { blocked.contains($0.note.pubkey) }) else { return }
        collected = collected.filter { !blocked.contains($0.value.note.pubkey) }
        reels = reels.filter { !blocked.contains($0.note.pubkey) }
    }

    // MARK: - Fetching

    private func fetch() {
        disconnect()

        var authors: [String]?
        if scope == .following {
            let follows = FeedService.shared.followedPubkeys
            guard !follows.isEmpty else {
                isLoading = false
                isLoadingMore = false
                followSetIsEmpty = true
                return
            }
            authors = follows
        }

        var filters: [[String: Any]] = []
        for stream in Stream.allCases where !exhausted.contains(stream) {
            var filter: [String: Any] = stream == .video
                ? ["kinds": Self.videoKinds, "limit": 100]
                : ["kinds": [1], "limit": 300]
            if let authors { filter["authors"] = authors }
            if let cursor = cursors[stream] { filter["until"] = cursor }
            filters.append(filter)
        }
        guard !filters.isEmpty else {
            isLoading = false
            isLoadingMore = false
            return
        }

        let relayURLs = Self.relayURLs(scope: scope)
        guard !relayURLs.isEmpty else {
            isLoading = false
            isLoadingMore = false
            loadFailed = true
            return
        }

        let subId = "reels-\(UUID().uuidString.prefix(8))"
        let blocked = ConfigService.shared.activeAccountBlockedHexPubkeys
        countBeforeFetch = collected.count
        eoseThisFetch = 0
        pageOldest.removeAll()

        for (relayIndex, url) in relayURLs.enumerated() {
            let client = WebSocketClient()
            client.isTemporary = true
            clients.append(client)

            client.messageSubject
                .receive(on: DispatchQueue.main)
                .sink { [weak self] message in
                    self?.handle(message: message, relay: relayIndex, blocked: blocked)
                }
                .store(in: &cancellables)

            client.connect(url: url)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
                let req = ["REQ", subId] + filters.map { $0 as Any }
                if let data = try? JSONSerialization.data(withJSONObject: req),
                   let text = String(data: data, encoding: .utf8) {
                    client.send(text: text)
                }
            }
        }

        loadTimeout?.invalidate()
        loadTimeout = Timer.scheduledTimer(withTimeInterval: 10.0, repeats: false) { [weak self] _ in
            Task { @MainActor [weak self] in self?.finishPage() }
        }
    }

    /// The device's own relay (and its feed cache) answer first and hold the
    /// follow set's history; the external relays fill in the rest.
    static func relayURLs(scope: RecipeScope) -> [URL] {
        var strings: [String] = []
        if scope == .following,
           RelayProcessManager.shared.isRunning,
           !RelayProcessManager.shared.isBooting {
            let local = ConfigService.shared.config.nostrURL
            strings += [local, local + "/feed"]
        }
        let configured = ConfigService.shared.config.activeFeedRelays
        strings += configured.isEmpty ? ["wss://relay.primal.net", "wss://nos.lol"] : configured
        var seen = Set<String>()
        return strings.filter { seen.insert($0).inserted }.compactMap { URL(string: $0) }
    }

    // MARK: - Private

    private func handle(message: String, relay: Int, blocked: Set<String>) {
        guard let data = message.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [Any],
              let type = json.first as? String else { return }

        if type == "EOSE" {
            eoseThisFetch += 1
            // Every relay has answered — no reason to sit out the timeout.
            if eoseThisFetch >= clients.count { finishPage() } else { schedulePublish(immediate: true) }
            return
        }

        guard type == "EVENT", json.count >= 3,
              let event = json[2] as? [String: Any],
              let id = event["id"] as? String,
              let pubkey = event["pubkey"] as? String,
              let createdAt = event["created_at"] as? Int64,
              let kind = event["kind"] as? Int,
              let tags = event["tags"] as? [[String]],
              let content = event["content"] as? String,
              kind == 1 || Self.videoKinds.contains(kind)
        else { return }

        let stream: Stream = kind == 1 ? .note : .video
        pageOldest[relay, default: [:]][stream] = min(pageOldest[relay]?[stream] ?? createdAt, createdAt)

        guard !blocked.contains(pubkey), collected[id] == nil else { return }

        let note = FeedNote(
            id: id,
            pubkey: pubkey,
            content: content,
            createdAt: Date(timeIntervalSince1970: TimeInterval(createdAt)),
            tags: tags,
            kind: kind
        )
        guard let reel = Reel(note: note, createdAt: createdAt),
              seenVideoURLs.insert(reel.videoURL.absoluteString).inserted
        else { return }

        collected[id] = reel
        schedulePublish(immediate: false)
    }

    /// Events arrive one message at a time from several relays; batch them so
    /// the pager is not re-laid-out per event.
    private func schedulePublish(immediate: Bool) {
        publishWork?.cancel()
        let work = DispatchWorkItem { [weak self] in self?.publish() }
        publishWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + (immediate ? 0 : 0.35), execute: work)
    }

    private func publish() {
        if orderIsFrozen {
            let shown = Set(reels.map(\.id))
            let arrivals = collected.values
                .filter { !shown.contains($0.id) }
                .sorted(by: Reel.newestFirst)
            if !arrivals.isEmpty { reels.append(contentsOf: arrivals) }
        } else {
            reels = collected.values.sorted(by: Reel.newestFirst)
        }
    }

    private func finishPage() {
        guard isFetching else { return }
        publish()

        // No relay answered at all: a network failure, not the end of
        // history. Leave the cursors alone so a later swipe can retry.
        let gotAnything = !pageOldest.isEmpty
        if !gotAnything && eoseThisFetch == 0 {
            disconnect()
            isLoading = false
            isLoadingMore = false
            loadFailed = reels.isEmpty && !followSetIsEmpty
            return
        }

        // Next cursor per stream: the NEWEST of the relays' oldest events.
        // Taking the oldest overall would jump a dense relay past history a
        // sparse relay never had; this way every relay resumes where it
        // stopped, and the overlap is deduplicated by event id.
        for stream in Stream.allCases where !exhausted.contains(stream) {
            let oldest = pageOldest.values.compactMap { $0[stream] }
            if let resume = oldest.max() {
                cursors[stream] = resume - 1
            } else {
                exhausted.insert(stream)
            }
        }

        emptyPages = collected.count == countBeforeFetch ? emptyPages + 1 : 0
        disconnect()
        isLoading = false
        isLoadingMore = false
        loadFailed = false

        // Keep paging when the viewer is already near the end — a page of
        // notes with no video in it would otherwise park the pager on its last
        // reel with nothing left to trigger the next page.
        guard !reachedEnd else { return }
        let shownIndex = shownReelId.flatMap { id in reels.firstIndex { $0.id == id } } ?? 0
        if reels.isEmpty || shownIndex >= reels.count - 3 {
            if reels.isEmpty { isLoading = true } else { isLoadingMore = true }
            fetch()
        }
    }
}

/// One page of the Reels feed: a note and the single video it plays.
struct Reel: Identifiable, Equatable {
    let note: FeedNote
    let createdAt: Int64
    let videoURL: URL
    /// MIME from the event, for extensionless Blossom URLs.
    let mimeType: String?
    /// Poster frame the event publishes, shown until the video has a frame.
    let posterURL: URL?
    /// Width / height when the event says; otherwise learned from the player.
    let aspectRatio: CGFloat?
    let title: String?
    /// The note text with the video link itself taken out.
    let caption: String

    var id: String { note.id }

    static func == (lhs: Reel, rhs: Reel) -> Bool { lhs.id == rhs.id }

    static func newestFirst(_ a: Reel, _ b: Reel) -> Bool {
        if a.createdAt != b.createdAt { return a.createdAt > b.createdAt }
        return a.id > b.id
    }

    init?(note: FeedNote, createdAt: Int64) {
        // A note with a content warning autoplaying full-screen defeats the
        // warning, so those stay in the timeline where they can be blurred.
        if note.tags.contains(where: { $0.first == "content-warning" }) { return nil }

        let imeta = Self.imetaFields(note.tags)
        let isVideoEvent = ReelsFeedService.videoKinds.contains(note.kind)

        // First URL that is known to be video, in the order the note gives them.
        var pick: (url: URL, mime: String?)?
        for entry in imeta {
            guard let raw = entry["url"], let url = URL(string: raw), Self.isHTTP(url) else { continue }
            let mime = entry["m"]
            if let mime, mime.lowercased().hasPrefix("video/") { pick = (url, mime); break }
            if MediaKindResolver.cachedKind(for: url, mimeHint: mime) == .video { pick = (url, mime); break }
        }
        if pick == nil {
            for url in note.mediaURLs where Self.isHTTP(url) {
                if MediaKindResolver.cachedKind(for: url) == .video { pick = (url, nil); break }
            }
        }
        // Older NIP-71 events carry a bare `url` tag instead of imeta.
        if pick == nil, isVideoEvent,
           let raw = note.tags.first(where: { $0.count >= 2 && $0[0] == "url" })?[1],
           let url = URL(string: raw), Self.isHTTP(url) {
            let mime = note.tags.first(where: { $0.count >= 2 && $0[0] == "m" })?[1]
            pick = (url, mime)
        }
        guard let pick else { return nil }

        let entry = imeta.first { $0["url"] == pick.url.absoluteString }
        self.note = note
        self.createdAt = createdAt
        self.videoURL = pick.url
        self.mimeType = pick.mime
        self.posterURL = (entry?["image"] ?? entry?["thumb"]
            ?? note.tags.first(where: { $0.count >= 2 && ($0[0] == "image" || $0[0] == "thumb") })?[1])
            .flatMap { URL(string: $0) }
        self.aspectRatio = (entry?["dim"]).flatMap(Self.aspect(fromDim:))
        let title = note.tags.first(where: { $0.count >= 2 && $0[0] == "title" })?[1]
            .trimmingCharacters(in: .whitespacesAndNewlines)
        self.title = (title?.isEmpty ?? true) ? nil : title
        self.caption = note.content
            .replacingOccurrences(of: pick.url.absoluteString, with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func isHTTP(_ url: URL) -> Bool {
        let scheme = url.scheme?.lowercased()
        return scheme == "https" || scheme == "http"
    }

    /// Each NIP-92 `imeta` tag as a field dictionary (`url`, `m`, `dim`, `image`…).
    private static func imetaFields(_ tags: [[String]]) -> [[String: String]] {
        tags.compactMap { tag in
            guard tag.first == "imeta", tag.count >= 2 else { return nil }
            var fields: [String: String] = [:]
            for field in tag.dropFirst() {
                guard let space = field.firstIndex(of: " ") else { continue }
                let key = String(field[..<space])
                let value = field[field.index(after: space)...].trimmingCharacters(in: .whitespaces)
                // First value wins — `image` and `fallback` may repeat.
                if fields[key] == nil, !value.isEmpty { fields[key] = value }
            }
            return fields
        }
    }

    /// `dim 1080x1920` → 0.5625.
    static func aspect(fromDim dim: String) -> CGFloat? {
        let parts = dim.lowercased().split(separator: "x")
        guard parts.count == 2, let w = Double(parts[0]), let h = Double(parts[1]), w > 0, h > 0 else { return nil }
        return CGFloat(w / h)
    }
}
